cmaes/_safe_cma.py

```diff
@@ -0,0 +1,702 @@
+            / ((n_dim + 2) ** 2 + alpha_cov * mu_eff / 2),
+        )
+        assert c1 <= 1 - cmu, "invalid learning rate for the rank-one update"
+        assert cmu <= 1 - c1, "invalid learning rate for the rank-μ update"
+
+        cm = 1  # (eq. 54)
+
+        # learning rate for the cumulation for the step-size control (eq.55)
+        c_sigma = (mu_eff + 2) / (n_dim + mu_eff + 5)
+        d_sigma = 1 + 2 * max(0, math.sqrt((mu_eff - 1) / (n_dim + 1)) - 1) + c_sigma
+        assert (
+            c_sigma < 1
+        ), "invalid learning rate for cumulation for the step-size control"
+
+        # learning rate for cumulation for the rank-one update (eq.56)
+        cc = (4 + mu_eff / n_dim) / (n_dim + 4 + 2 * mu_eff / n_dim)
+        assert cc <= 1, "invalid learning rate for cumulation for the rank-one update"
+
+        self._n_dim = n_dim
+        self._popsize = population_size
+        self._mu = mu
+        self._mu_eff = mu_eff
+
+        self._cc = cc
+        self._c1 = c1
+        self._cmu = cmu
+        self._c_sigma = c_sigma
+        self._d_sigma = d_sigma
+        self._cm = cm
+
+        # E||N(0, I)|| (p.28)
+        self._chi_n = math.sqrt(self._n_dim) * (
+            1.0 - (1.0 / (4.0 * self._n_dim)) + 1.0 / (21.0 * (self._n_dim**2))
+        )
+
+        self._weights = weights
+
+        # evolution path
+        self._p_sigma = np.zeros(n_dim)
+        self._pc = np.zeros(n_dim)
+
+        if cov is None:
+            self._C = np.eye(n_dim)
+        else:
+            assert cov.shape == (n_dim, n_dim), "Invalid shape of covariance matrix"
+            self._C = cov
+
+        self._D: Optional[np.ndarray] = None
+        self._B: Optional[np.ndarray] = None
+
+        self._rng = np.random.RandomState(seed)
+
+        # initial distribution parameter
+        self._sigma = sigma
+        mean, sigma = self._init_distribution(sigma)
+
+        assert sigma > 0, "sigma must be non-zero positive value"
+
+        assert np.all(
+            np.abs(mean) < _MEAN_MAX
+        ), f"Abs of all elements of mean vector must be less than {_MEAN_MAX}"
+
+        self._mean = mean.copy()
+        self._sigma = sigma
+
+        # bounds contains low and high of each parameter.
+        assert bounds is None or _is_valid_bounds(bounds, mean), "invalid bounds"
+        self._bounds = bounds
+        self._n_max_resampling = n_max_resampling
+
+        self._g = 0
+
+        # Termination criteria
+        self._tolx = 1e-12 * sigma
+        self._tolxup = 1e4
+        self._tolfun = 1e-12
+        self._tolconditioncov = 1e14
+
+        self._funhist_term = 10 + math.ceil(30 * n_dim / population_size)
+        self._funhist_values = np.empty(self._funhist_term * 2)
+
+    def _compute_lipschitz_constant(self) -> np.ndarray:
+
+        likelihood = gpytorch.likelihoods.GaussianLikelihood(
+            noise_constraint=gpytorch.constraints.GreaterThan(0)
+        )
+        likelihood.noise = 0
```

Comment:
I'm curious about the validity of this setting because, as mentioned in this thread:
> Some small value, but don't make it too small or numerical performance will suffer. I recommend 1e-4.
https://github.com/cornellius-gp/gpytorch/issues/1196#issuecomment-651193968

This implies there's room for improvement in this setting, but I'm not sure for now, so let's see how it goes.